### ISSUE TITLE
Increasing the number of vehicles in Figure Eight scenario 

### ISSUE THREAD

**Issue by pnp91** [flow-project/flow#522]
**Choose an option**
This is a: 
- [x] Question
- [x] Feature request

**Describe the issue**
Figure eight scenario only accommodates a maximum of 14 vehicles in Flow. Even if I give more than this, it is not effective. But I am planning to extend this to 100 vehicles involving RL and human vehicles. I just want to know the reason behind this limitation to be able to apply in a real world scenario. Can I add another route with this and expand the capacity_ Like merging one figure eight and one ring wave scenario or merging two figure eight scenarios to increase its capacity. 

**Expected behavior**
Many vehicles in a single road track

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Version [run python -c "import flow; print(flow.__version__)"]: 
 Leave empty if not applicable.

**Additional context**
https://flow.readthedocs.io/en/latest/examples.html#figure-eight

**Comment by eugenevinitsky** [comment]
Yes, you can absolutely combine multiple scenarios or increase the length of the roadway to accommodate more vehicles. If you point me at the specific scenario we can try and figure out the exact change you would need to make.

**Comment by pnp91** [comment]
@eugenevinitsky Thank you so much for your kind reply and I deeply regret and apologize for delay in my response. My main goal is to accommodate more autonomous vehicles in figure eight scenario. And not to restrict to 14 vehicles. A solution for this, I have identified to merge the scenarios inorder to increase the capacity of ring. I will definitely try increasing this and get back to  you in-case I face further difficulties.

**Comment by eugenevinitsky** [comment]
Sounds good! You should also be able to increase the ring length parameter, which should give you more space to stick extra vehicles in.

**Comment by pnp91** [comment]
Oh yes I am aware of this parameter but wasn't aware that it can hold extra vehicles in. Thank you so much for the solution. That sounds a good option. And especially thanks a lot for timely constant support and encouragement, explaining us with all patience without ignoring. I have noticed it many times and feel so happy for this pro-active support from Team Flow. Kudos!!

**Comment by jingyanping** [comment]
When I was running the  flow/examples/rllib/stabilizing_the_ring.py. it generated a series of files, such as 'checkpoint_20',  'checkpoint_40'..., 'progress.csv',  'result.json'.  When I finished training, I didn't know how to look at my training results. I read the documentation find that have two ways to resolve that,

 First : input code `python ./visualizer_rllib.py /ray_results/result_dir 1` at terminal, but this command is reported an error can't open file....

 Second:` tensorboard --logdir=~/ray_results`, the terminal always displays "TensorBoard. 1.9.0 at http://ubuntu:6006 (Press CTRL+C to quit)"
 So I want to know how I should look at the results after my training.@pnp91@eugenevinitsky

**Comment by pnp91** [comment]
@jingyanping The first command is used to visualize the files from any particular checkpoint number. But for this, you have to be in the path of Visualize file. You can make "cd flow/flow/visualize" and then view the output at one particular checkpoint. For example if you wish to see the output at checkpoint 40, then type the following after you are at visualize folder 
python ./visualizer_rllib.py /home/<your path>/ray_results/<exp_tag(here for this example, it is "lord_of_numrings">/<folder name inside this exp_tag><give space><number of checkpoint. Here it is"40">
In the above command, I have used < > for you to be editable and do not use with these brackets.

Second : It is for visualizing a graph for training steps, rewards, entropy and so on.. You can press on the link to open in a new tab and check the graph of results. 

Hope it helps you.
@eugenevinitsky : As I was also mentioned in this question and I know the answer, I have replied. I hope it should not be a problem.

**Comment by jingyanping** [comment]
Thank you very much for your kind reply, i followed your suggestion,
'''
(flow) sean@ubuntu:~/flow/flow/visualize$ python ./visualizer_rllib.py /ray_results/stabilizing_the_ring/checkpoint 40
'''
But it has reported an error:
'''
Traceback (most recent call last):
File "./visualizer_rllib.py", line 384, in <module>
visualizer_rllib(args)
File "./visualizer_rllib.py", line 57, in visualizer_rllib
config = get_rllib_config(result_dir)
File "/home/sean/flow/flow/utils/rllib.py", line 169, in get_rllib_config
"Could not find params.json in either the checkpoint dir or "
ValueError: Could not find params.json in either the checkpoint dir or its parent directory.
'''

But in my ray_results, i find the file params.json. Can you give me some advice?

At 2019-07-18 15:24:46, "niyathipriya" <notifications@github.com> wrote:


@jingyanping The first command is used to visualize the files from any particular checkpoint number. But for this, you have to be in the path of Visualize file. You can make "cd flow/flow/visualize" and then view the output at one particular checkpoint. For example if you wish to see the output at checkpoint 40, then type the following after you are at visualize folder
python ./visualizer_rllib.py /home//ray_results/<exp_tag(here for this example, it is "lord_of_numrings">/<number of checkpoint. Here it is"40">
In the above command, I have used < > for you to be editable and do not use with these brackets.

Second : It is for visualizing a graph for training steps, rewards, entropy and so on.. You can press on the link to open in a new tab and check the graph of results.

Hope it helps you.
@eugenevinitsky : As I was also mentioned in this question and I know the answer, I have replied. I hope it should not be a problem.

**Comment by pnp91** [comment]
I think the problem is you used wrong name for <exp_tag>.. It is not the name of .py file. But there is a parameter inside .py file called exp_tag, you have to copy that name. It is also the same name which you have inside your "ray_results" folder. 
One more problem is checkpoint number. You just need to mention the number and not as "checkpoint_40".
Hopefully my below example screenshot helps you.
![Screenshot from 2019-07-18 09-56-05](https://user-images.githubusercontent.com/33332697/61440013-e1a2f680-a942-11e9-88c0-51a746d45881.png)

**Comment by jingyanping** [comment]
oh yes, i resolve that problem follow  your advice,Thank you very much!!!

**Comment by pnp91** [comment]
Cool! 😀

Kindly contact me in case of any concerns.


Thanks & Regards,
Niyathipriya Pasupuleti
Tel : +4915143598348

**Comment by jingyanping** [comment]
I'm sorry to disturb you again.
About the "tensorboard-logdir=~/ray_results" command, Is it meaning that when i open one terminal to train the vehicle, and input "tensorboard-logdir=~/ray_results" in another terminal at the same time? If i do that, the terminal show that" TensorBoard 1.9. 0 at http://ubuntu:6006 (Press CTRL+C to quit)", but i can't open the Link, i think that is not right.

At 2019-07-18 15:24:46, "niyathipriya" <notifications@github.com> wrote:


@jingyanping The first command is used to visualize the files from any particular checkpoint number. But for this, you have to be in the path of Visualize file. You can make "cd flow/flow/visualize" and then view the output at one particular checkpoint. For example if you wish to see the output at checkpoint 40, then type the following after you are at visualize folder
python ./visualizer_rllib.py /home//ray_results/<exp_tag(here for this example, it is "lord_of_numrings">/<number of checkpoint. Here it is"40">
In the above command, I have used < > for you to be editable and do not use with these brackets.

Second : It is for visualizing a graph for training steps, rewards, entropy and so on.. You can press on the link to open in a new tab and check the graph of results.

Hope it helps you.
@eugenevinitsky : As I was also mentioned in this question and I know the answer, I have replied. I hope it should not be a problem.

**Comment by pnp91** [comment]
no.. you can do in any terminal after the training is completed.. you can
right click on the link and open it in new tab.
[image: Screenshot from 2019-07-18 14-00-38.png]

**Comment by jingyanping** [comment]
Thanks!!!
i will try it!

At 2019-07-18 20:02:36, "niyathipriya" <notifications@github.com> wrote:
no.. you can do in any terminal after the training is completed.. you can
right click on the link and open it in new tab.
[image: Screenshot from 2019-07-18 14-00-38.png]

**Comment by jingyanping** [comment]
hello，
I'm sorry trouble you again. i have trained the auto car in a closed network using PPO algorithm, it can dissipate stop-and-go waves in the closed loop.(The PPO algorithm can also be replaced with other algorithm,such as A3C,DQN,DDPG,right?). i trained the auto car in closed network, Can it be migrated to an open-loop network,for example, on a single-lane road with one or two ramps? If so,can you give me some advice about how to do that?
sincerely
jing

At 2019-07-18 16:00:40, "niyathipriya" <notifications@github.com> wrote:


I think the problem is you used wrong name for <exp_tag>.. It is not the name of .py file. But there is a parameter inside .py file called exp_tag, you have to copy that name. It is also the same name which you have inside your "ray_results" folder.
One more problem is checkpoint number. You just need to mention the number and not as "checkpoint_40".
Hopefully my below example screenshot helps you.

**Comment by pnp91** [comment]
Hello Ping,

yes you can do that. You need to change in the scenario which you define
with the specify routes, specify connections and specify edges functions.
This is well explained further in tutorial on "how do we create our own
custom scenarios". Although building a scenario is  not a very easy task, I
would suggest you to follow the tutorial given in FLOW. They have
appreciable description which helps us to build our own tracks. Hope this
helps.

**Comment by jingyanping** [comment]
OK,thank you very much for your reply!



At 2019-07-25 16:05:24, "niyathipriya" <notifications@github.com> wrote:
Hello Ping,

yes you can do that. You need to change in the scenario which you define
with the specify routes, specify connections and specify edges functions.
This is well explained further in tutorial on "how do we create our own
custom scenarios". Although building a scenario is not a very easy task, I
would suggest you to follow the tutorial given in FLOW. They have
appreciable description which helps us to build our own tracks. Hope this
helps.

**Comment by pnp91** [comment]
No problem :) All the best

**Comment by eugenevinitsky** [comment]
Hi Jing, this is actually a very good question and something that's not incredibly easy to do at the moment (though we're working on making it easier)! It's definitely possible, but you have to mess around a good deal with rllib to make it work.

**Comment by jingyanping** [comment]
Hi eugenevinitsky，I read a paper(Dissipating stop-and-go waves in the closed and open networks via deep reinforcement learning) from your team, which is to transfer the experience gained in the closed network to the open network through transfer learning. But i cant't understand how this is done now.
Another problem is that the auto car i trained in the closed network used the PPO algorithm in tutorial 03, i am learning PPO algorithm recently, i want to know if the PPO algorithm code is in ray/python/ray/rllib/agents/ppo/ppo.py? Because i don't understand the training process and reward mechanism of the vehicle, i want to thoroughly understand how the vehicle is trained, just observe the speed and  headway of the front car?

**Comment by jingyanping** [comment]
Dear Niyathipriya，

Hello，I am so sorry to disturb you, I have encountered a problem and I would like to ask you. In the experiment of using the ‘AV’ car to dissipate the loop stop-and-go wave, in ‘stabilizing_the_ring.py’, when reading the relevant paper, the MLP strategy is selected for the agent. but I did not find the MLP strategy related program in the provided program, but in 'ray/ Python/ray/rllib/models' found programs like fcnet.py, so I was confused. Mentioned in the paper 'Flow:Architecture and Benchmarking for Reinforcement Learning in Traffic Control': "The most experiments,a  diagonal Gaussian MLP policy is used with hidden layers (100,50,25) and tanh nonlinearity,The expeiment stabilizing the ring,uses a hidden layer of shape(3,3)..."
sincerely
Jing




在 2019-07-18 16:35:05，"niyathipriya" <notifications@github.com> 写道：
Cool!

Kindly contact me in case of any concerns.


Thanks & Regards,
Niyathipriya Pasupuleti
Tel : +4915143598348
